File: 029_ReverseLinkedList.cpp
// Topic: Linked List
// Reverse Linked List: Reverse the order of nodes in a singly linked list.
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* reverseListIter(ListNode* head)
{
    ListNode* prev = nullptr;
    while (head)
    {
        ListNode* nxt = head->next;
        head->next = prev;
        prev = head;
        head = nxt;
    }
    return prev;
}

ListNode* reverseListRec(ListNode* head)
{
    if (!head || !head->next)
        return head;
    ListNode* p = reverseListRec(head->next);
    head->next->next = head;
    head->next = nullptr;
    return p;
}

void printList(ListNode* head)
{
    while (head)
    {
        cout << head->val;
        if (head->next)
            cout << "->";
        head = head->next;
    }
    cout << '\n';
}

int main()
{
    ListNode* a = new ListNode(1);
    a->next = new ListNode(2);
    a->next->next = new ListNode(3);
    a->next->next->next = new ListNode(4);
    cout << "orig: ";
    printList(a);
    auto r = reverseListIter(a);
    cout << "iter: ";
    printList(r);

    // reverse back with recursion
    auto r2 = reverseListRec(r);
    cout << "rec:  ";
    printList(r2);
    return 0;
}

File: 029_ReverseLinkedList.py
"""
Topic: Linked List
Reverse Linked List: iterative and recursive solutions
Time Complexity: O(n)
Space Complexity: O(1) iterative, O(n) recursive
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def reverse_list_iter(head: Optional[ListNode]) -> Optional[ListNode]:
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev

def reverse_list_rec(head: Optional[ListNode]) -> Optional[ListNode]:
    if not head or not head.next:
        return head
    p = reverse_list_rec(head.next)
    head.next.next = head
    head.next = None
    return p

def make_list(vals):
    dummy = ListNode(0)
    cur = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next

def to_list(head):
    res = []
    while head:
        res.append(head.val); head = head.next
    return res

if __name__ == '__main__':
    h = make_list([1,2,3,4,5])
    print(to_list(reverse_list_iter(h)))  # [5,4,3,2,1]
    h2 = make_list([1,2])
    print(to_list(reverse_list_rec(h2)))  # [2,1]
